// ngx-countup/esm2020/lib/count-up.module.mjs
import { NgModule } from '@angular/core';
import { CountUpDirective } from './count-up.directive';
import * as i0 from "@angular/core";
export class CountUpModule {
}
CountUpModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: CountUpModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
CountUpModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: CountUpModule, declarations: [CountUpDirective], exports: [CountUpDirective] });
CountUpModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: CountUpModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.2.2", ngImport: i0, type: CountUpModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [],
                    declarations: [CountUpDirective],
                    exports: [CountUpDirective]
                }]
        }] });
