// @siemens/ngx-datatable/esm2022/lib/directives/disable-row.directive.mjs
import { Directive, Input } from '@angular/core';
import * as i0 from "@angular/core";
/**
 * Row Disable Directive
 * Use this to disable/enable all children elements
 * Usage:
 *  To disable
 * 		<div [disabled]="true" disable-row >
 * 		</div>
 *  To enable
 *  	<div [disabled]="false" disable-row >
 * 		</div>
 */
export class DisableRowDirective {
    set disabled(val) {
        this._disabled = val;
        if (val) {
            this.disableAllElements();
        }
    }
    ;
    get disabled() {
        return this._disabled;
    }
    constructor(element) {
        this.element = element;
        this._disabled = false;
    }
    disableAllElements() {
        const el = this.element?.nativeElement;
        if (!el) {
            return;
        }
        Array.from(el.querySelectorAll('*')).forEach(child => {
            child?.setAttribute('disabled', '');
        });
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.6", ngImport: i0, type: DisableRowDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "17.0.6", type: DisableRowDirective, selector: "[disable-row]", inputs: { disabled: "disabled" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.6", ngImport: i0, type: DisableRowDirective, decorators: [{
            type: Directive,
            args: [{ selector: '[disable-row]' }]
        }], ctorParameters: () => [{ type: i0.ElementRef }], propDecorators: { disabled: [{
                type: Input
            }] } });
