// lucide-angular/esm2020/lib/lucide-angular.module.mjs
import { NgModule, Optional } from '@angular/core';
import { LucideAngularComponent } from './lucide-angular.component';
import { LUCIDE_ICONS, LucideIconProvider } from './lucide-icon.provider';
import { Icons } from './icons.provider';
import * as i0 from "@angular/core";
const legacyIconProviderFactory = (icons) => {
    return new LucideIconProvider(icons ?? {});
};
export class LucideAngularModule {
    static pick(icons) {
        return {
            ngModule: LucideAngularModule,
            providers: [
                { provide: LUCIDE_ICONS, multi: true, useValue: new LucideIconProvider(icons) },
                {
                    provide: LUCIDE_ICONS,
                    multi: true,
                    useFactory: legacyIconProviderFactory,
                    deps: [[new Optional(), Icons]],
                },
            ],
        };
    }
}
LucideAngularModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: LucideAngularModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
LucideAngularModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: LucideAngularModule, declarations: [LucideAngularComponent], exports: [LucideAngularComponent] });
LucideAngularModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: LucideAngularModule, imports: [[]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.3.12", ngImport: i0, type: LucideAngularModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [LucideAngularComponent],
                    imports: [],
                    exports: [LucideAngularComponent],
                }]
        }] });
