// angular-ng-stepper/esm2020/lib/ng-stepper.module.mjs
import { NgModule } from '@angular/core';
import { NgStepperComponent } from './ng-stepper.component';
import { CdkStepperModule } from '@angular/cdk/stepper';
import { CommonModule } from '@angular/common';
import * as i0 from "@angular/core";
export class NgStepperModule {
}
NgStepperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: NgStepperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgStepperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.5", ngImport: i0, type: NgStepperModule, declarations: [NgStepperComponent], imports: [CommonModule,
        CdkStepperModule], exports: [NgStepperComponent] });
NgStepperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: NgStepperModule, imports: [CommonModule,
        CdkStepperModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.5", ngImport: i0, type: NgStepperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgStepperComponent],
                    imports: [
                        CommonModule,
                        CdkStepperModule
                    ],
                    exports: [NgStepperComponent]
                }]
        }] });
